ote_sdk/ote_sdk/utils/argument_checks.py

```diff
@@ -0,0 +1,217 @@
+    if isinstance(expected_type, typing._GenericAlias):  # type: ignore
+        origin_class = expected_type.__dict__.get("__origin__")
+        # Checking origin class
+        raise_value_error_if_parameter_has_unexpected_type(
+            parameter=parameter,
+            parameter_name=parameter_name,
+            expected_type=origin_class,
+        )
+        # Checking nested elements
+        if issubclass(origin_class, typing.Sequence):
+            check_nested_elements_type(
+                iterable=parameter,
+                parameter_name=parameter_name,
+                expected_type=expected_type.__dict__.get("__args__")[0],
+            )
+        elif origin_class == dict:
+            key, value = expected_type.__dict__.get("__args__")
+            if value == float:
+                value = (int, float, floating)
+            check_dictionary_keys_values_type(
+                parameter=parameter,
+                parameter_name=parameter_name,
+                expected_key_class=key,
+                expected_value_class=value,
+            )
+    else:
+        raise_value_error_if_parameter_has_unexpected_type(
+            parameter=parameter,
+            parameter_name=parameter_name,
+            expected_type=expected_type,
+        )
+
+
+def check_required_parameters_type(parameter_name_expected_type: list):
+    """
+    Function raises ValueError exception if required parameters have unexpected type
+    :param parameter_name_expected_type: list with tuples that contain parameter, name for exception message and
+    expected type
+    """
+    for parameter, name, expected_type in parameter_name_expected_type:
+        check_parameter_type(
+            parameter=parameter, parameter_name=name, expected_type=expected_type
+        )
+
+
+def check_optional_parameters_type(parameter_name_expected_type: list):
+    """
+    Function checks if optional parameters exist and raises ValueError exception if one of them has unexpected type
+    :param parameter_name_expected_type: list with tuples that contain optional parameter, name for exception message
+    and expected type
+    """
+    for parameter, name, expected_type in parameter_name_expected_type:
+        if parameter is not None:
+            check_parameter_type(
+                parameter=parameter, parameter_name=name, expected_type=expected_type
+            )
+
+
+def check_required_and_optional_parameters_type(
+    required_parameters: list, optional_parameters: list
+):
+    """Function raises ValueError exception if required or optional parameter has unexpected type"""
+    check_required_parameters_type(required_parameters)
+    check_optional_parameters_type(optional_parameters)
+
+
+def check_nested_elements_type(iterable, parameter_name, expected_type):
+    """Function raises ValueError exception if one of elements in collection has unexpected type"""
+    for element in iterable:
+        raise_value_error_if_parameter_has_unexpected_type(
```

Comment:
I would use here `check_parameter_type` -- it will be a sort of recursion.
And `parameter_name` in the call I would set to `f"element in {parameter_name}"` -- it will allow to work with `List[List[int]]`.